alacritty/src/renderer/mod.rs

```diff
@@ -691,37 +672,7 @@ impl QuadRenderer {
 
     /// Draw all rectangles simultaneously to prevent excessive program swaps.
     pub fn draw_rects(&mut self, props: &SizeInfo, rects: Vec<RenderRect>) {
-        // Swap to rectangle rendering program.
-        unsafe {
-            // Swap program.
-            gl::UseProgram(self.rect_program.id);
-
-            // Remove padding from viewport.
-            gl::Viewport(0, 0, props.width() as i32, props.height() as i32);
-
-            // Change blending strategy.
-            gl::BlendFuncSeparate(gl::SRC_ALPHA, gl::ONE_MINUS_SRC_ALPHA, gl::SRC_ALPHA, gl::ONE);
-
-            // Setup data and buffers.
-            gl::BindVertexArray(self.rect_vao);
-            gl::BindBuffer(gl::ARRAY_BUFFER, self.rect_vbo);
-
-            // Position.
-            gl::VertexAttribPointer(
-                0,
-                2,
-                gl::FLOAT,
-                gl::FALSE,
-                (size_of::<f32>() * 2) as _,
-                ptr::null(),
-            );
-            gl::EnableVertexAttribArray(0);
-        }
-
-        // Draw all the rects.
-        for rect in rects {
-            self.render_rect(&rect, props);
-        }
+        self.rect_renderer.draw(&self.rect_program, props, rects);
 
         // Deactivate rectangle program again.
```

Comment:
If the program is activated inside the rect renderer, shouldn't the reset be performed inside it as well?

I'm just talking about stuff like `gl::UseProgram(0)`. Being enabled inside the draw call and enabled outside seems a bit strange, but I can see why you wouldn't want to reset things that the rect renderer technically has no knowledge of.

Reply:
I've moved state strictly related to rects (buffers, program) inside the rect_renderer, and all other reasonably-common state (viewport, blend mode) to outside.